test/test_population.py

```diff
@@ -0,0 +1,77 @@
+import copy
+import gp
+import pytest
+import numpy as np
+
+
+def test_assert_mutation_rate(rng_seed, genome_params, mutation_rate):
+    with pytest.raises(ValueError):
+        gp.Population(5, -0.1, rng_seed, genome_params)
+
+    with pytest.raises(ValueError):
+        gp.Population(5, 1.1, rng_seed, genome_params)
+
+    # assert that no error is thrown for a suitable mutation rate
+    gp.Population(5, mutation_rate, rng_seed, genome_params)
+
+
+def test_init_random_parent_population(population_params, genome_params):
+    pop = gp.Population(**population_params, genome_params=genome_params)
+    assert len(pop.parents) == population_params["n_parents"]
+
+
+def test_champion(population_params, genome_params):
+    pop = gp.Population(**population_params, genome_params=genome_params)
+    for i, parent in enumerate(pop.parents):
+        parent.fitness = float(i)
+    assert pop.champion == pop.parents[-1]
+
+
+def test_crossover_too_small(population_simple_fitness):
+    pop = population_simple_fitness
+    # Breeding pool too small
+    with pytest.raises(AssertionError):
+        pop.crossover(pop.parents[:1], 2)
+
+
+def test_crossover_one_offspring_all_parents(population_simple_fitness):
+    pop = population_simple_fitness
+    # Check is best parent is chosen if n_offsprings = 1
+    offspring = pop.crossover(pop.parents, 1)
+    assert offspring[0] == pop.champion
+
+
+def test_crossover_one_offspring_breeding_pool(population_simple_fitness):
+    pop = population_simple_fitness
+    # Check is best parent in smaller breeding pool is chosen if n_offsprings = 1
+    offspring = pop.crossover(pop.parents[:3], 1)
```

Comment:
why first three?

Reply:
Just picking something less than the full list of parents. Do you have a preference for a different magic number?

Reply:
no, not really. maybe add a comment? although i find it hard to think of a comment that is not redundant